imagetest/testworkflow.go

```diff
@@ -434,3 +451,25 @@ func parseResult(res testResult) *testSuite {
 	ret.Name = fmt.Sprintf("%s-%s", res.testWorkflow.Name, res.testWorkflow.ShortImage)
 	return ret
 }
+
+func (t *TestWorkflow) lastResolveStep() (*daisy.Step, error) {
+	reverseDependencies := make(map[string]string)
+	for dependent, dependencies := range t.wf.Dependencies {
+		for _, dependency := range dependencies {
+			if _, ok := reverseDependencies[dependency]; ok {
+				return nil, fmt.Errorf("dependency %s is not linear", dependency)
+			}
+			reverseDependencies[dependency] = dependent
+		}
+	}
+
+	lastStepName := resolveStep(reverseDependencies, createDisksStepName)
+	return t.wf.Steps[lastStepName], nil
+}
+
+func resolveStep(reverseDependencies map[string]string, currentStep string) string {
+	if _, ok := reverseDependencies[currentStep]; !ok {
+		return currentStep
+	}
+	return resolveStep(reverseDependencies, reverseDependencies[currentStep])
+}
```

Comment:
there are a few issues with this algorithm.

first, you will never be able to create the reverse-dependency map on any workflow that has more than one VM, because the wait-$VMNAME steps all depend on the create-vms step.

second, you are looking for the last step of the whole workflow, but we need the last step for a given VM. third, using recursion this way has some risky rough edges. you can get infinite recursion in some scenarios and you've got no safeguards here for it. here is a lengthy proposed alternative:

```suggestion
func  (t *TestWorkflow) getLastStepForVM(vmname string) (*daisy.Step, error) {
	step := "wait-" + vmname
	if _, ok := t.wf.Dependencies[step]; !ok {
		return nil, fmt.Errorf("no step " + step)
	}
	rdeps := make(map[string][]string)
	for dependent, dependencies := range t.wf.Dependencies {
		for _, dependency := range dependencies {
			rdeps[dependency] = append(rdeps[dependency], dependent)
		}
	}

	for {
		deps, ok := rdeps[step]
		if !ok {
			// no more steps depend on this one
			break
		}
		if len(deps) > 1 {
			return nil, fmt.Errorf("Workflow has non-linear dependencies!")
		}
		step = deps[0]
	}

	return t.wf.Steps[step], nil

}

```

Reply:
Good catch. To work for more than one vm, should start specific VM which could be start at `wait-vm` step